### PR TITLE
[JUJU-1129] Refactor state resources persistence code to remove abstractions

### PR DESCRIPTION
**Pull request by wallyworld**
Remove a lot of unnecessary abstractions from the resources state code.
The interface to api callers is the same.

A lot of the new LOC is rewriting the state unit tests to drive mongo directly. Previously stubs were used and these didn't give good coverage.

## QA steps

Deploy the dummy-resource charm
Attach a new resource
run juju list-resources
run resource-get hook command and inspect charm resources dir

### CODE REVIEW

File: state/application.go [hunk 1]
```diff
@@ -619,15 +619,8 @@ func (op *DestroyApplicationOperation) unassignBranchOps() ([]txn.Op, error) {
...
-	persist, err := st.ResourcesPersistence()
-	if errors.IsNotSupported(err) {
-		// Nothing to see here, move along.
-		return nil, nil
-	}
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	ops, err := persist.NewRemoveResourcesOps(applicationID)
+	resources := st.Resources().(*resourcePersistence)
```
Comment (hpidcock) [inline state/application.go:622]: is this guaranteed to be *resourcePersistence?

File: state/application.go [hunk 2]
```diff
@@ -1493,11 +1486,8 @@ func incCharmModifiedVersionOps(applicationID string) []txn.Op {
 
 func (a *Application) resolveResourceOps(resourceIDs map[string]string) ([]txn.Op, error) {
 	// Collect pending resource resolution operations.
-	resources, err := a.st.Resources()
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	return resources.NewResolvePendingResourcesOps(a.doc.Name, resourceIDs)
+	resources := a.st.Resources().(*resourcePersistence)
```
Comment (hpidcock) [inline state/application.go:1489]: actually maybe just add an unexported resources method that returns *resourcePersistence

**Review by hpidcock — APPROVED** [review]

File: state/application.go (hunk 1 above)
Comment (wallyworld) [reply state/application.go:622]: yes, i could raise an error just in case

File: state/application.go (hunk 2 above)
Comment (wallyworld) [reply state/application.go:1489]: ah good idea

**Pull request merged**